alembic/versions/20231204_d3cdbea3d43b_marc_export_cleanup_migration.py

```diff
@@ -0,0 +1,150 @@
+"""MARC Export cleanup migration.
+
+Revision ID: d3cdbea3d43b
+Revises: 0039f3f12014
+Create Date: 2023-12-04 17:23:26.396526+00:00
+
+"""
+from typing import Optional
+from urllib.parse import urlparse
+
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+from alembic import op
+from core.migration.util import migration_logger
+from core.service.container import container_instance
+
+# revision identifiers, used by Alembic.
+revision = "d3cdbea3d43b"
+down_revision = "0039f3f12014"
+branch_labels = None
+depends_on = None
+
+
+def parse_key_from_url(url: str, bucket: str) -> Optional[str]:
+    """Parse the key from a URL.
+
+    :param url: The URL to parse.
+    :return: The key, or None if the URL is not a valid S3 URL.
+    """
+    parsed_url = urlparse(url)
+
+    if f"/{bucket}/" in parsed_url.path:
+        return parsed_url.path.split(f"/{bucket}/", 1)[1]
+
+    if bucket in parsed_url.netloc:
+        return parsed_url.path.lstrip("/")
+
+    return None
+
+
+def upgrade() -> None:
+    # Before removing the cachedmarcfiles table, we need to delete the cachedmarcfiles
+    # from s3.
+    services = container_instance()
+    public_s3 = services.storage.public()
+    log = migration_logger(revision)
+
+    # Check if there are any cachedmarcfiles in s3
+    connection = op.get_bind()
+    cached_files = connection.execute(
+        "SELECT r.mirror_url FROM cachedmarcfiles cmf JOIN representations r ON cmf.representation_id = r.id"
+    ).all()
+    if public_s3 is None and len(cached_files) > 0:
+        raise RuntimeError(
+            "There are cachedmarcfiles in the database, but no public s3 storage configured!"
+        )
```

Comment:
👍🏻 yes I agree. I didn't think about that case. 

I think we still want to fail the migration, since in production we want to make sure this case doesn't happen, or at least be intentional about it if it does.

Maybe I'll split this migration into two, one that is a pure database migration, and one that does the S3 part. So that one can be skipped if necessary in dev.

Reply:
Okay I pushed a commit that splits up the migration to make this easier to deal with in dev.